compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -1123,7 +1130,7 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
     val body1 = typed(tree.body, pt1)
     typr.println(i"typed bind $tree pt = $pt1 bodytpe = ${body1.tpe}")
     body1 match {
-      case UnApply(fn, Nil, arg :: Nil) if tree.body.isInstanceOf[untpd.Typed] =>
+      case UnApply(fn, Nil, arg :: Nil) if tree.body.isInstanceOf[untpd.Typed] && !body1.tpe.isError =>
```

Comment:
This is now too restrictive, it means that the following does not typecheck:
```scala
  def isFoo(x: Any) = x match {
    case x @ ExtractFoo() =>
      val b: Foo = x
      true
```
Because at this point we have:
```scala
tree.body = Apply(Ident(ExtractFoo),List())
```

(Note that explicitly writing `x @ (_: Foo)` does not work either because then `tree.body` is an instance of `untpd.Parens`, not `untpd.Typed`)